File: running-app-backend/social/serializers/post.py
from rest_framework import serializers

from social.models import EventPost,\
                        ClubPost, \
                        EventPostLike, \
                        ClubPostLike

from account.serializers.user_abbr import UserAbbrSerializer 

from social.serializers.post_comment import ClubPostCommentSerializer, \
                                            EventPostCommentSerializer
from social.serializers.post_like import ClubPostLikeSerializer, \
                                        EventPostLikeSerializer

from utils.pagination import CommonPagination

class ClubPostSerializer(serializers.ModelSerializer):
    user = UserAbbrSerializer()
    likes = serializers.SerializerMethodField()
    check_user_like = serializers.SerializerMethodField()

    def get_check_user_like(self, instance):
        user_id = self.context["user"].id
        post_id = instance.id
        instance =  ClubPostLike.objects\
            .filter(user_id=user_id, post_id=post_id).first()
        return instance.id if instance else None

    def get_likes(self, instance):
        queryset = instance.likes.all()
        # paginator = CommonPagination(page_size=100)
        # paginated_queryset = paginator.paginate_queryset(queryset, self.context["request"])
        # return UserAbbrSerializer(paginated_queryset, many=True, read_only=True).data
        return UserAbbrSerializer(queryset, many=True, read_only=True, context={
            "user": self.context["user"],
            "request": self.context["request"],
        }).data
    
    class Meta:
        model = ClubPost
        # exclude = ("likes", "user")
        fields = (
            "id", "user", "total_likes", "likes", "total_comments",
            "title", "content", "created_at", "check_user_like"
        )
        read_only_fields = ("id", "created_at")

class DetailClubPostSerializer(serializers.ModelSerializer):
    user = UserAbbrSerializer()
    likes = serializers.SerializerMethodField()
    comments = serializers.SerializerMethodField()
    check_user_like = serializers.SerializerMethodField()

    def get_check_user_like(self, instance):
        user_id = self.context["user"].id
        post_id = instance.id
        instance =  ClubPostLike.objects\
            .filter(user_id=user_id, post_id=post_id).first()
        return instance.id if instance else None
    
    def get_likes(self, instance):
        queryset = instance.likes.all()
        paginator = CommonPagination(page_size=100)
        paginated_queryset = paginator.paginate_queryset(queryset, self.context["request"])
        return UserAbbrSerializer(paginated_queryset, many=True, read_only=True, context={
            "user": self.context["user"],
            "request": self.context["request"],
        }).data
    
    def get_comments(self, instance):
        queryset = instance.comments.all()
        paginator = CommonPagination(page_size=5, page_query_param="cmt_pg")
        paginated_queryset = paginator.paginate_queryset(queryset, self.context["request"])
        return ClubPostCommentSerializer(
            paginated_queryset, many=True, read_only=True, context={
                "request": self.context["request"]
            }).data
        
    class Meta:
        model = ClubPost
        fields = (
            "id", "user", "total_likes", "total_comments", "likes", "comments",
            "title", "content", "created_at", "check_user_like"
        )
        read_only_fields = ("created_at", "id", "author_id")

class CreateClubPostSerializer(serializers.ModelSerializer):
    user_id = serializers.UUIDField(write_only=True)
    club_id = serializers.UUIDField(write_only=True)
    user = serializers.SerializerMethodField()
    
    def get_user(self, instance):
        return UserAbbrSerializer(
            instance.user,
            context={
                "request": self.context["request"],
            }    
        ).data
    
    def create(self, validated_data):
        user_id = validated_data.pop("user_id")
        club_id = validated_data.pop("club_id")
        post = ClubPost.objects.create(
            user_id=user_id, club_id=club_id, **validated_data)
        return post
    
    class Meta:
        model = ClubPost
        fields = (
            "id", "user", "title", "content",
            "created_at", "updated_at", "club_id", "user_id", "total_likes", "total_comments",
        )
        extra_kwargs = {
            "id": {"read_only": True},
            "post": {"read_only": True},
            "user": {"read_only": True},
            "total_likes": {"read_only": True},
            "total_comments": {"read_only": True},
            # "user_id": {"write_only": True},
        }

class UpdateClubPostSerializer(serializers.ModelSerializer):
    user = serializers.SerializerMethodField()

    def get_user(self, instance):
        return UserAbbrSerializer(
            instance.user,
            context={
                "request": self.context["request"],
            }    
        ).data
    
    class Meta:
        model = ClubPost
        fields = (
            "id", "user", "created_at", "updated_at", 
            "total_likes", "total_comments",
            "title", "content"
        )
        read_only_fields = (
            "id", "user",
            "created_at", "updated_at", 
            "total_likes", "total_comments"
        )

class EventPostSerializer(serializers.ModelSerializer):
    user = UserAbbrSerializer()
    likes = serializers.SerializerMethodField()
    check_user_like = serializers.SerializerMethodField()

    def get_check_user_like(self, instance):
        user_id = self.context["user"].id
        post_id = instance.id
        instance =  EventPostLike.objects\
            .filter(user_id=user_id, post_id=post_id).first()
        return instance.id if instance else None
    
    def get_likes(self, instance):
        queryset = instance.likes.all()
        # paginator = CommonPagination(page_size=100)
        # paginated_queryset = paginator.paginate_queryset(queryset, self.context["request"])
        # return UserAbbrSerializer(paginated_queryset, many=True, read_only=True).data
        return UserAbbrSerializer(queryset, many=True, read_only=True, context={
            "user": self.context["user"],
            "request": self.context.get("request"),
        }).data
    class Meta:
        model = ClubPost
        # exclude = ("likes", "user")
        fields = (
            "id", "user", "total_likes", "likes", "total_comments",
            "title", "content", "created_at", "updated_at", "check_user_like")
        read_only_fields = ("id", "created_at")

class DetailEventPostSerializer(serializers.ModelSerializer):
    user = UserAbbrSerializer()
    likes = serializers.SerializerMethodField()
    comments = serializers.SerializerMethodField()
    check_user_like = serializers.SerializerMethodField()

    def get_check_user_like(self, instance):
        user_id = self.context["user"].id
        post_id = instance.id
        instance =  EventPostLike.objects\
            .filter(user_id=user_id, post_id=post_id).first()
        return instance.id if instance else None
    
    def get_likes(self, instance):
        queryset = instance.likes.all()
        paginator = CommonPagination(page_size=100)
        paginated_queryset = paginator.paginate_queryset(queryset, self.context["request"])
        return UserAbbrSerializer(paginated_queryset, many=True, read_only=True, context={
            "user": self.context["user"],
            "request": self.context["request"],
        }).data
    
    def get_comments(self, instance):
        queryset = instance.comments.all()
        paginator = CommonPagination(page_size=5, page_query_param="cmt_pg")
        paginated_queryset = paginator.paginate_queryset(queryset, self.context["request"])
        return EventPostCommentSerializer(
            paginated_queryset, many=True, read_only=True, context={
                "request": self.context["request"]
            }).data
    
    class Meta:
        model = EventPost
        fields = (
            "id", "user", "total_likes", "total_comments", "likes", "comments",
            "title", "content", "created_at", "updated_at", "check_user_like")
        read_only_fields = ("created_at", "id", "author_id")
        

class CreateEventPostSerializer(serializers.ModelSerializer):
    user_id = serializers.UUIDField(write_only=True)
    event_id = serializers.UUIDField(write_only=True)
    user = serializers.SerializerMethodField()

    def get_user(self, instance):
        return UserAbbrSerializer(
            instance.user,
            context={
                "request": self.context["request"],
            }    
        ).data
    
    def create(self, validated_data):
        user_id = validated_data.pop("user_id")
        event_id = validated_data.pop("event_id")
        post = EventPost.objects.create(
            user_id=user_id, event_id=event_id, **validated_data)
        return post
    
    class Meta:
        model = EventPost
        fields = (
            "id", "user", "title", "content",
            "created_at", "updated_at", "event_id", "user_id", "total_likes", "total_comments",
        )
        extra_kwargs = {
            "id": {"read_only": True},
            "post": {"read_only": True},
            "user": {"read_only": True},
            "total_likes": {"read_only": True},
            "total_comments": {"read_only": True},
            # "user_id": {"write_only": True},
        }

class UpdateEventPostSerializer(serializers.ModelSerializer):
    user = serializers.SerializerMethodField()

    def get_user(self, instance):
        return UserAbbrSerializer(
            instance.user,
            context={
                "request": self.context["request"],
            }
        ).data
    
    class Meta:
        model = EventPost
        fields = (
            "id", "user", "created_at", "updated_at", 
            "total_likes", "total_comments",
            "title", "content"
        )
        read_only_fields = (
            "id", "user",
            "created_at", "updated_at", 
            "total_likes", "total_comments"
        )

File: running_app/lib/models/social/post.dart

import 'package:running_app/models/account/activity.dart';
import 'package:running_app/models/account/author.dart';
import 'package:running_app/models/account/like.dart';
import 'package:running_app/models/activity/club.dart';
import 'package:running_app/models/activity/event.dart';
import 'package:running_app/models/social/post_comment.dart';

class Post {
  final String? id;
  final Author? user;
  final String? title;
  final String? content;
  final String? createdAt;
  final String? updatedAt;
  String? checkUserLike;

  Post({
    required this.id,
    required this.user,
    required this.title,
    required this.content,
    required this.createdAt,
    required this.updatedAt,
    required this.checkUserLike,
  });

  Post.fromJson(Map<String, dynamic> json)
      : id = json['id'],
        user = Author.fromJson(json['user']),
        title = json['title'],
        content = json['content'],
        createdAt = json['created_at'],
        updatedAt = json['updated_at'],
        checkUserLike = json['check_user_like'];

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'user': user,
      'title': title,
      'content': content,
      'created_at': createdAt,
    };
  }

  @override
  String toString() {
    return "${toJson()}";
  }
}

class ClubPost extends Post {
  int totalLikes;
  int totalComments;
  final List<Like>? likes;

  void increaseTotalLikes() {
    totalLikes += 1;
  }

  void increaseTotalComments() {
    totalComments += 1;
  }

  void decreaseTotalLikes() {
    totalLikes -= 1;
  }

  void decreaseTotalComments() {
    totalComments -= 1;
  }

  ClubPost({
    String? id,
    Author? user,
    String? title,
    String? content,
    String? updatedAt,
    String? createdAt,
    String? checkUserLike,
    required this.likes,
    required this.totalLikes,
    required this.totalComments,
  }) : super(
      id: id,
      user: user,
      title: title,
      content: content,
      createdAt: createdAt,
      updatedAt: updatedAt,
      checkUserLike: checkUserLike
  );

  ClubPost.fromJson(Map<String, dynamic> json)
      : likes = (json['likes'] as List<dynamic>)
        .map((likeJson) => Like.fromJson(likeJson))
        .toList(),
        totalLikes = json['total_likes'],
        totalComments = json['total_comments'],
        super.fromJson(json);

  @override
  String toString() {
    return '$title by $totalLikes in $totalComments';
  }
}

class DetailClubPost extends ClubPost {
  final List<ClubPostComment>? comments;

  DetailClubPost({
    String? id,
    String? title,
    String? content,
    String? createdAt,
    List<Like>? likes,
    required int totalLikes,
    required int totalComments,
    Author? user,
    required this.comments,
  }) : super(
      id: id,
      title: title,
      content: content,
      createdAt: createdAt,
      totalLikes: totalLikes,
      totalComments: totalComments,
      user: user,
      likes: likes
  );

  DetailClubPost.fromJson(Map<String, dynamic> json)
      : comments = (json['comments'] as List<dynamic>)
            .map((commentJson) => ClubPostComment.fromJson(commentJson))
            .toList(),
        super.fromJson(json);

  @override
  String toString() {
    return '$title by $user $likes $comments';
  }
}

class EventPost extends Post {
  int totalLikes;
  int totalComments;
  final List<Like>? likes;

  void increaseTotalLikes() {
    totalLikes += 1;
  }

  void increaseTotalComments() {
    totalComments += 1;
  }

  void decreaseTotalLikes() {
    totalLikes -= 1;
  }

  void decreaseTotalComments() {
    totalComments -= 1;
  }

  EventPost({
    String? id,
    Author? user,
    String? title,
    String? content,
    String? createdAt,
    String? updatedAt,
    String? checkUserLike,
    required this.likes,
    required this.totalLikes,
    required this.totalComments,
  }) : super(
      id: id,
      user: user,
      title: title,
      content: content,
      createdAt: createdAt,
      updatedAt: updatedAt,
      checkUserLike: checkUserLike
  );

  EventPost.fromJson(Map<String, dynamic> json)
      : likes = (json['likes'] as List<dynamic>?)
          ?.map((likeJson) => Like.fromJson(likeJson))
          .toList(),
        totalLikes = json['total_likes'],
        totalComments = json['total_comments'],
        super.fromJson(json);

  @override
  String toString() {
    return '$title by $totalLikes in $totalComments';
  }
}

class DetailEventPost extends EventPost {
  final List<EventPostComment>? comments;

  DetailEventPost({
    String? id,
    String? title,
    String? content,
    String? createdAt,
    List<Like>? likes,
    required int totalLikes,
    required int totalComments,
    Author? user,
    required this.comments,
  }) : super(
      id: id,
      title: title,
      content: content,
      createdAt: createdAt,
      totalLikes: totalLikes,
      totalComments: totalComments,
      user: user,
      likes: likes
  );

  DetailEventPost.fromJson(Map<String, dynamic> json)
      : comments = (json['comments'] as List<dynamic>?)
          ?.map((commentJson) => EventPostComment.fromJson(commentJson))
          .toList(),
        super.fromJson(json);

  @override
  String toString() {
    return '$title by $user $likes $comments';
  }
}

class CreateClubPost {
  final String? title;
  final String? content;
  final String? userId;
  final String? clubId;

  CreateClubPost({
    required this.userId,
    required this.clubId,
    required this.title,
    required this.content,
  });

  Map<String, dynamic> toJson() {
    return {
      'user_id': userId,
      'club_id': clubId,
      'title': title,
      'content': content,
    };
  }

  @override
  String toString() {
    return "${toJson()}";
  }
}

class CreateEventPost {
  final String? title;
  final String? content;
  final String? userId;
  final String? eventId;

  CreateEventPost({
    required this.userId,
    required this.eventId,
    required this.title,
    required this.content,
  });

  Map<String, dynamic> toJson() {
    return {
      'user_id': userId,
      'event_id': eventId,
      'title': title,
      'content': content,
    };
  }

  @override
  String toString() {
    return "${toJson()}";
  }
}


class UpdatePost {
  final String? title;
  final String? content;

  UpdatePost({
    required this.title,
    required this.content,
  });

  Map<String, dynamic> toJson() {
    return {
      'title': title,
      'content': content,
    };
  }

  @override
  String toString() {
    return "${toJson()}";
  }
}


